system/Helpers/filesystem_helper.php

```diff
@@ -91,7 +91,9 @@ function directory_mirror(string $originDir, string $targetDir, bool $overwrite
             $target = $targetDir . substr($origin, $dirLen);
 
             if ($file->isDir()) {
-                mkdir($target, 0755);
+                if (! is_dir($target)) {
```

Comment:
Can you make this not nested if? I mean you can combine this with the outer `if` by a `&&`

Reply:
```diff
--- a/system/Helpers/filesystem_helper.php
+++ b/system/Helpers/filesystem_helper.php
@@ -90,10 +90,8 @@ if (! function_exists('directory_mirror')) {
             $origin = $file->getPathname();
             $target = $targetDir . substr($origin, $dirLen);
 
-            if ($file->isDir()) {
-                if (! is_dir($target)) {
-                    mkdir($target, 0755);
-                }
+            if ($file->isDir() && ! is_dir($target)) {
+                mkdir($target, 0755);
             } elseif (! is_file($target) || ($overwrite && is_file($target))) {
                 copy($origin, $target);
             }
```
```sh-session
$ ./phpunit tests/system/Helpers/FilesystemHelperTest.php
PHPUnit 9.5.10 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.0.14
Configuration: /Users/kenji/work/codeigniter/CodeIgniter4/phpunit.xml

.......E.............................                             37 / 37 (100%)

Time: 00:00.256, Memory: 12.00 MB

There was 1 error:

1) CodeIgniter\Helpers\FilesystemHelperTest::testDirectoryMirrorSkipExistingFolder
ErrorException: copy(): The first argument to copy() function cannot be a directory

/Users/kenji/work/codeigniter/CodeIgniter4/system/Helpers/filesystem_helper.php:96
/Users/kenji/work/codeigniter/CodeIgniter4/tests/system/Helpers/FilesystemHelperTest.php:181

ERRORS!
Tests: 37, Assertions: 87, Errors: 1.
```

Reply:
@paulbalandan If you want to remove if in if, this is the code.
But I'm not sure it is correct.

```diff
--- a/system/Helpers/filesystem_helper.php
+++ b/system/Helpers/filesystem_helper.php
@@ -90,10 +90,12 @@ if (! function_exists('directory_mirror')) {
             $origin = $file->getPathname();
             $target = $targetDir . substr($origin, $dirLen);
 
-            if ($file->isDir()) {
-                if (! is_dir($target)) {
-                    mkdir($target, 0755);
-                }
+            if ($file->isDir() && ! is_dir($target)) {
+                mkdir($target, 0755);
+            } elseif ($file->isDir() && is_dir($target)) {
+                continue;
+            } elseif (is_dir($target)) {
+                continue;
             } elseif (! is_file($target) || ($overwrite && is_file($target))) {
                 copy($origin, $target);
             }
```

Reply:
How about this:
```diff
diff --git a/system/Helpers/filesystem_helper.php b/system/Helpers/filesystem_helper.php
index 89d4d507f..104b739a3 100644
--- a/system/Helpers/filesystem_helper.php
+++ b/system/Helpers/filesystem_helper.php
@@ -90,11 +90,11 @@ if (! function_exists('directory_mirror')) {
             $origin = $file->getPathname();
             $target = $targetDir . substr($origin, $dirLen);

-            if ($file->isDir()) {
-                if (! is_dir($target)) {
-                    mkdir($target, 0755);
-                }
-            } elseif (! is_file($target) || ($overwrite && is_file($target))) {
+            if ($file->isDir() && ! is_dir($target)) {
+                mkdir($target, 0755);
+            }
+
+            if ($file->isFile() && (! is_file($target) || $overwrite)) {
                 copy($origin, $target);
             }
         }
```

Reply:
Changing the conditions of the if statements does not seem to make it easier to understand.